src/hotspot/share/runtime/vmStructs.hpp

```diff
@@ -188,13 +188,18 @@ class VMStructs {
 #ifdef ASSERT
 
 // This macro checks the type of a VMStructEntry by comparing pointer types
-#define CHECK_NONSTATIC_VM_STRUCT_ENTRY(typeName, fieldName, type)                 \
- {typeName *dummyObj = NULL; type* dummy = &dummyObj->fieldName;                   \
-  assert(offset_of(typeName, fieldName) < sizeof(typeName), "Illegal nonstatic struct entry, field offset too large"); }
+#define CHECK_NONSTATIC_VM_STRUCT_ENTRY(typeName, fieldName, type) {               \
+    char space[sizeof (typeName)];                                                 \
+    typeName *dummyObj = (typeName *)space; type* dummy = &dummyObj->fieldName;    \
+    assert(offset_of(typeName, fieldName) < sizeof(typeName), "Illegal nonstatic struct entry, field offset too large"); \
```

Comment:
I think what this assert is really trying to do is verify that fieldName designates an ordinary data member
rather than a static data member.  That's probably possible to determine via some metaprogramming
rather than using `offset_of`.  I'll see if I can come up with something.